File: LeetCode/Easy/0876_Middle_of_Linked_List/Solution.py
from typing import *
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

## 2021-12-27 (fast-slow)
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        count = 1
        fast = head
        slow = head
        
        while fast is not None:
            if count % 2 == 0:
                slow = slow.next
            fast = fast.next
            count += 1
        
        return slow

File: LeetCode/Easy/0876_Middle_of_Linked_List/Solution.cpp
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

// 2022-12-04 (fast-slow)
class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        /*
            input: a linked list
            output: the middle node of the list (2nd middle if even number)

            approach:
            - fast and slow pointer starting at a ghost node
            - fast pointer moves forward 2 every time slow pointer moves 1
            - stop iterating when fast pointer == null
            - return slow pointer
        */
        ListNode *ghost = new ListNode(-1, head);
        ListNode *fast = ghost, *slow = ghost;
        while (fast != nullptr) {
            slow = slow->next;
            fast = fast->next;
            if (fast != nullptr)
                fast = fast->next;
        }
        return slow;

    }
};